src/lib/timing-src-connector.js

```diff
@@ -0,0 +1,217 @@
+ *
+ * Proposed spec: https://webtiming.github.io/timingobject/
+ * V3 Spec: https://timingsrc.readthedocs.io/en/latest/
+ * Demuxed talk: https://www.youtube.com/watch?v=cZSjDaGDmX8
+ *
+ * This class makes it easy to connect Vimeo.Player to a provided TimingObject via Vimeo.Player.setTimingSrc(myTimingObject, options) and the synchronization will be handled automatically.
+ *
+ * There are 5 general responsibilities in TimingSrcConnector:
+ *
+ * 1. `updatePlayer()` which sets the player's currentTime, playbackRate and pause/play state based on current state of the TimingObject.
+ * 2. `updateTimingObject()` which sets the TimingObject's position and velocity from the player's state.
+ * 3. `playerUpdater` which listens for change events on the TimingObject and will respond by calling updatePlayer.
+ * 4. `timingObjectUpdater` which listens to the player events of seeked, play and pause and will respond by calling `updateTimingObject()`.
+ * 5. `maintainPlaybackPosition` this is code that constantly monitors the player to make sure it's always in sync with the TimingObject. This is needed because videos will generally not play with precise time accuracy and there will be some drift which becomes more noticeable over longer periods (as noted in the timing-object spec). More details on this method below.
+ */
+export class TimingSrcConnector extends EventTarget {
+    logger;
+
+    /**
+     * @param {PlayerControls} player
+     * @param {TimingObject} timingObject
+     * @param {TimingSrcConnectorOptions} options
+     * @param {Logger} logger
+     */
+    constructor(player, timingObject, options = {}, logger) {
+        super();
+        this.logger = logger;
+        this.init(timingObject, player, { ...defaultOptions, ...options });
+    }
+
+    disconnect() {
+        this.dispatchEvent(new Event('disconnect'));
+    }
+
+
+    /**
+     * @param {TimingObject} timingObject
+     * @param {PlayerControls} player
+     * @param {TimingSrcConnectorOptions} options
+     * @return {Promise<void>}
+     */
+    async init(timingObject, player, options) {
+        await this.waitForTOReadyState(timingObject, 'open');
+
+        if (options.role === 'viewer') {
+            await this.updatePlayer(timingObject, player, options);
+            const playerUpdater = subscribe(timingObject, 'change', () => this.updatePlayer(timingObject, player, options));
+            const positionSync = this.maintainPlaybackPosition(timingObject, player, options);
+            this.addEventListener('disconnect', () => {
+                positionSync.cancel();
+                playerUpdater.cancel();
+            });
+        }
+        else {
+            await this.updateTimingObject(timingObject, player);
+            const timingObjectUpdater = subscribe(player, ['seeked', 'play', 'pause', 'ratechange'], () => this.updateTimingObject(timingObject, player), 'on', 'off');
```

Comment:
I feel like this `subscribe` function might be more readable when used if it took an object of K/Vs rather than a long list of arguments that don't really mean anything unless you're familiar with the function's definition.

Reply:
I'd leave it up to you. Personally, I prefer the readability of the second approach.